[doorbird] Doorbird Binding initial contribution

mhilbush opened the pull request:
This is a binding for Doorbird video doorbells.

Features include:
- support for models D101 and D210x
- channels for doorbell pressed and motion detected events, including switch channels, timestamp channels, and image channels
- image channel that can be refreshed on a user defined number of seconds
- image and timestamp channels for historical images (up to last 50 images) and timestamps for button press and motion detected events
- history image “montage” channels for displaying a montage (photo strip) of the last n doorbell press and/or motion history images
- channels for triggering open door and light relays

Forum discussion is [here](https://community.openhab.org/t/doorbird-video-doorbell-binding/73564).

This binding is ready for review.

Signed-off-by: Mark Hilbush <mark@hilbush.com>

ppieczul commented on bundles/org.openhab.binding.doorbird/src/main/java/org/openhab/binding/doorbird/internal/DoorbirdUdpListener.java:
```diff
@@ -0,0 +1,151 @@
+        receivePackets();
+    }
+
+    public void shutdown() {
```
I think there is a racing possible on socket so that main thread will have disposed of it and this thread will still use it. Maybe an exception will handle it, but it is not a graceful shutdown. This thread could handle interrupted exception and check for interrupted status instead of for socket to be null and close the socket. Main thread could only cancel(true) the job.

mhilbush replied:
I'll look into this.

mhilbush replied:
I remember now... Calling `cancel` won't interrupt the receive call on the UDP socket. We ran into this a while back on the [miio binding](https://community.openhab.org/t/xiaomi-robot-vacuum-binding/31317/462?u=mhilbush). If you still think there's an issue here, we can discuss further.

https://stackoverflow.com/questions/4670664/interrupt-a-thread-in-datagramsocket-receive

https://stackoverflow.com/questions/40055175/how-to-interrupt-a-blocking-call-to-udp-sockets-receive

https://stackoverflow.com/questions/7962121/how-can-i-stop-the-block-method-datagramsocket-receive-in-a-thread

ppieczul replied:
Oh, great to learn new things. I think if the only reason for IOException is closing the thread, this is good. But if IOException can occur for another reason, which I don't know, the racing is still possible and maybe a critical section on access to socket would be needed. Additionally, if it occurred, the thread would just wind down silently.

ppieczul commented on bundles/org.openhab.binding.doorbird/src/main/java/org/openhab/binding/doorbird/internal/DoorbirdUdpListener.java:
```diff
@@ -0,0 +1,151 @@
+        }
+    }
+
+    private boolean isDuplicate(DatagramPacket packet) {
```
Why would we receive duplicate packets? Does doorbird send them?

mhilbush replied:
Yes, it sends at least 2 packets per user defined in the Doorbird app. So, for example, if you have 3 users defined in the app, you'll get at least 6 packets.

ppieczul commented on bundles/org.openhab.binding.doorbird/README.md:
```diff
@@ -0,0 +1,217 @@
+
+| Channel ID | Item Type | Description |
+|------------|-----------|-------------|
+| doorbell | Switch | Updated to ON whenever doorbell is pressed |
```
I think this could be a trigger channel (not sure if they are still available in openHAB, they were year ago).
This would simplify the code, no need for delayed setting state to Off. Same for motion trigger.

mhilbush replied:
This is a good idea. As there are a number of people using the binding, I'll post this for discussion in the forum to get some feedback.

mhilbush replied:
This is a good idea. Since there are a number of people using the binding already, to ensure backwards compatibility, I could make this a trigger channel, then use a profile to retain the ability to link a Switch item to the channel. WDYT?

ppieczul replied:
I just learned and read about profiles and it looks like a great idea

mhilbush replied:
Ok, I'll give that a go. I've not used profiles before, so this will be a good opportunity to learn how to use them.

mhilbush replied:
I've implemented the trigger and profile on the doorbell and motion channels. I'm pretty happy with how it's working, so I'm going to push the change. I'm also planning to leave the jobs that set the state to OFF. At least for the `motion` channel, that's one of the few ways one can mitigate the Doorbird's hyperactive motion sensor...

mhilbush commented on bundles/org.openhab.binding.doorbird/README.md:
```diff
@@ -0,0 +1,217 @@
+| motion | Switch | Updated to ON when motion is detected |
+| motionTimestamp | DateTime | Timestamp when motion sensor was triggered |
+| motionImage | Image | Image captured when motion was detected |
+| light | Switc | Activates the light relay |
```
Thanks.

mhilbush commented on bundles/org.openhab.binding.doorbird/src/main/java/org/openhab/binding/doorbird/internal/DoorbirdEvent.java:
```diff
@@ -0,0 +1,255 @@
+        isDoorbellEvent = false;
+
+        int length = p.getLength();
+        byte[] data = Arrays.copyOf(p.getData(), length);
```
I don't recall. Could be left over from some changes to that section of code. I'll take a look.

mhilbush replied:
I think I did this to ensure I was working on a local copy of the data contained in the UDP packet.

mhilbush commented on bundles/org.openhab.binding.doorbird/src/main/java/org/openhab/binding/doorbird/internal/DoorbirdHandler.java:
```diff
@@ -0,0 +1,619 @@
+                updateState(CHANNEL_DOORBELL_TIMESTAMP, getLocalDateTimeType(dbImage.getTimestamp()));
+            }
+            updateState(CHANNEL_DOORBELL, OnOffType.OFF);
+        }, 0, TimeUnit.SECONDS);
```
My understanding is that calls to `handleCommand` should return quickly to avoid tying up threads from the ThingHandler threadpool.

mhilbush commented on bundles/org.openhab.binding.doorbird/src/main/java/org/openhab/binding/doorbird/internal/DoorbirdHandler.java:
```diff
@@ -0,0 +1,619 @@
+            return;
+        }
+        int value = ((DecimalType) command).intValue();
+        if (value < 0 || value > 50) {
```
The Doorbird backend service holds a maximum of the last 50 doorbell and motion images. I probably should make note of this in the code.

J-N-K commented on bundles/org.openhab.binding.doorbird/src/main/java/org/openhab/binding/doorbird/internal/DoorbirdHandler.java:
```diff
@@ -0,0 +1,633 @@
+
+    private void refreshDoorbellImageFromHistory() {
+        logger.debug("Handler: REFRESH doorbell image channel using most recent doorbell history image");
+        scheduler.schedule(() -> {
```
```suggestion
        scheduler.execute(() -> {
```

also similar calls

mhilbush replied:
I'll fix all these. This was left over from when I had an initial delay on these. I had since remove the delay but failed to change to `execute`.

J-N-K commented on bundles/org.openhab.binding.doorbird/src/main/java/org/openhab/binding/doorbird/internal/DoorbirdHandler.java:
```diff
@@ -0,0 +1,633 @@
+                updateState(CHANNEL_DOORBELL_TIMESTAMP, getLocalDateTimeType(dbImage.getTimestamp()));
+            }
+            updateState(CHANNEL_DOORBELL, OnOffType.OFF);
+        }, 0, TimeUnit.SECONDS);
```
```suggestion
        });
```

mhilbush replied:
Fixed

J-N-K commented on bundles/org.openhab.binding.doorbird/src/main/java/org/openhab/binding/doorbird/internal/DoorbirdHandlerFactory.java:
```diff
@@ -0,0 +1,73 @@
+@NonNullByDefault
+@Component(configurationPid = "binding.doorbird", service = ThingHandlerFactory.class)
+public class DoorbirdHandlerFactory extends BaseThingHandlerFactory {
+    private @NonNullByDefault({}) TimeZoneProvider timeZoneProvider;
```
Have a look at constructor injection (e.g. SNMP binding). You can remove the getters, setters and the annotations here.

mhilbush replied:
I converted to use constructor injection, but I'm having trouble building the binding now. It seems to be having an issue with `HttpClient`.

```
[ERROR] Failed to execute goal org.apache.karaf.tooling:karaf-maven-plugin:4.2.6:verify (karaf-feature-verification) on project org.openhab.binding.doorbird: Feature resolution failed for [openhab-binding-doorbird/2.5.0.SNAPSHOT]
[ERROR] Message: Unable to resolve root: missing requirement [root] osgi.identity; osgi.identity=openhab-binding-doorbird; type=karaf.feature; version=2.5.0.SNAPSHOT; filter:="(&(osgi.identity=openhab-binding-doorbird)(type=karaf.feature)(version>=2.5.0.SNAPSHOT))" [caused by: Unable to resolve openhab-binding-doorbird/2.5.0.SNAPSHOT: missing requirement [openhab-binding-doorbird/2.5.0.SNAPSHOT] osgi.identity; osgi.identity=org.openhab.binding.doorbird; type=osgi.bundle; version="[2.5.0.201908231958,2.5.0.201908231958]"; resolution:=mandatory [caused by: Unable to resolve org.openhab.binding.doorbird/2.5.0.201908231958: missing requirement [org.openhab.binding.doorbird/2.5.0.201908231958] osgi.service; filter:="(objectClass=org.eclipse.jetty.client.HttpClient)"; effective:=active]]
[
```

This is the relevant part of the handler factory code. Am I missing something obvious?

```
@NonNullByDefault
@Component(configurationPid = "binding.doorbird", service = ThingHandlerFactory.class)
public class DoorbirdHandlerFactory extends BaseThingHandlerFactory {
    private final TimeZoneProvider timeZoneProvider;
    private final HttpClient httpClient;

    @Activate
    public DoorbirdHandlerFactory(@Reference TimeZoneProvider timeZoneProvider, @Reference HttpClient httpClient) {
        this.timeZoneProvider = timeZoneProvider;
        this.httpClient = httpClient;
    }
```

mhilbush replied:
Nevermind. I got it sorted out. This is what I needed.

```
    @Activate
    public DoorbirdHandlerFactory(@Reference TimeZoneProvider timeZoneProvider,
            @Reference HttpClientFactory httpClientFactory) {
        this.timeZoneProvider = timeZoneProvider;
        this.httpClient = httpClientFactory.getCommonHttpClient();
    }
```

J-N-K commented on bundles/org.openhab.binding.doorbird/src/main/java/org/openhab/binding/doorbird/internal/listener/DoorbirdEvent.java:
```diff
@@ -0,0 +1,255 @@
+
+        static final @Nullable LazySodiumJava LAZY_SODIUM_JAVA_INSTANCE = loadLazySodiumJava();
+
+        static @Nullable LazySodiumJava loadLazySodiumJava() {
```
```suggestion
        private static @Nullable LazySodiumJava loadLazySodiumJava() {
```

mhilbush replied:
Fixed

J-N-K commented on bundles/org.openhab.binding.doorbird/src/main/java/org/openhab/binding/doorbird/internal/profile/DoorbirdProfileFactory.java:
```diff
@@ -0,0 +1,66 @@
+@NonNullByDefault
+@Component(service = { ProfileFactory.class, ProfileTypeProvider.class })
+public class DoorbirdProfileFactory implements ProfileFactory, ProfileTypeProvider {
+    private static final Set<ProfileType> SUPPORTED_PROFILE_TYPES = Stream.of(DoorbirdProfiles.DOORBELL_COMMAND_TYPE)
```
```suggestion
    private static final Set<ProfileType> SUPPORTED_PROFILE_TYPES = Collections.singleton(DoorbirdProfiles.DOORBELL_COMMAND_TYPE);
```

mhilbush replied:
Fixed

J-N-K commented on bundles/org.openhab.binding.doorbird/src/main/resources/ESH-INF/config/config.xml:
```diff
@@ -0,0 +1,53 @@
+		</parameter>
+		<parameter name="userId" type="text" required="true">
+			<label>User ID</label>
+			<description>User ID</description>
```
remove if same as label

mhilbush replied:
Removed

J-N-K commented on bundles/org.openhab.binding.doorbird/src/main/resources/ESH-INF/config/config.xml:
```diff
@@ -0,0 +1,53 @@
+		</parameter>
+		<parameter name="userPassword" type="text" required="true">
+			<label>Password</label>
+			<description>Password</description>
```
remove if same as label

mhilbush replied:
Removed

J-N-K commented on bundles/org.openhab.binding.doorbird/src/main/resources/ESH-INF/config/config.xml:
```diff
@@ -0,0 +1,53 @@
+			<required>true</required>
+			<context>password</context>
+		</parameter>
+		<parameter name="imageRefreshRate" type="integer" min="2" max="600" required="false">
```
```suggestion
		<parameter name="imageRefreshRate" type="integer" min="2" max="600">
```

thats the default anyway

mhilbush replied:
Fixed all occurrences

J-N-K commented on bundles/org.openhab.binding.doorbird/src/main/resources/ESH-INF/config/config.xml:
```diff
@@ -0,0 +1,53 @@
+		<parameter name="doorbirdId" type="text" required="true">
+			<label>ID</label>
+			<description>Doorbird Identifier (e.g. 1)</description>
+			<required>true</required>
```
remove (already set above)

mhilbush replied:
Fixed all occurrences

J-N-K commented on bundles/org.openhab.binding.doorbird/src/main/resources/ESH-INF/config/config.xml:
```diff
@@ -0,0 +1,53 @@
+		<parameter name="doorbirdHost" type="text" required="true">
+			<label>Host</label>
+			<description>Hostname or IP address of Doorbird</description>
+			<required>true</required>
```
remove (already set above)

mhilbush replied:
Removed

J-N-K commented on bundles/org.openhab.binding.doorbird/src/main/resources/ESH-INF/config/config.xml:
```diff
@@ -0,0 +1,53 @@
+		<parameter name="userPassword" type="text" required="true">
+			<label>Password</label>
+			<description>Password</description>
+			<required>true</required>
```
remove (already set above)

mhilbush replied:
Removed

J-N-K commented on bundles/org.openhab.binding.doorbird/NOTICE:
```diff
@@ -0,0 +1,25 @@
+
+== Third-party Content
+
+lazysodium-java
```
If we do not include but only use the libraries you can remove that.

mhilbush replied:
Ah, ok. I didn't understand the distinction. Removed.

J-N-K commented on bundles/org.openhab.binding.doorbird/README.md:
```diff
@@ -0,0 +1,239 @@
+
+| Parameter | Required/Optional | Description |
+|-----------|-------------------|-------------|
+| ID | Required | The ID number of the Doorbird device. This is usually a single digit (e.g. 1). |
```
Please check formatting

mhilbush replied:
This does format correctly (see here https://github.com/mhilbush/openhab2-addons/tree/doorbird-new/bundles/org.openhab.binding.doorbird#thing-configuration).

And it seems to have the benefit of not being tedious to maintain with all the extra white space.

J-N-K replied:
I think there are different opinions :-). I prefer it aligned because it's easier to see in which columns the text is. AFAIK there is no strict guildeline.

J-N-K commented on bundles/org.openhab.binding.doorbird/src/main/java/org/openhab/binding/doorbird/internal/DoorbirdHandler.java:
```diff
@@ -0,0 +1,633 @@
+    private final Logger logger = LoggerFactory.getLogger(DoorbirdHandler.class);
+
+    // Get a dedicated threadpool for the listener thread
+    private final ScheduledExecutorService doorbirdScheduler = ThreadPoolManager.getScheduledPool("doorbirdHandler");
```
Can you comment why https://www.eclipse.org/smarthome/documentation/development/guidelines.html#d-runtime-behavior #2 does not apply here?

mhilbush replied:
For permanent threads, such as this thread that listens continuously for events from the device, it's recommended to not permanently tie up threads from the common scheduler pool.

J-N-K replied:
Ok for me.

mhilbush commented on bundles/org.openhab.binding.doorbird/pom.xml:
```diff
@@ -0,0 +1,38 @@
+  </dependencies>
+
+  <properties>
+    <bnd.importpackage>
```
Removed

mhilbush commented on bundles/org.openhab.binding.doorbird/src/main/java/org/openhab/binding/doorbird/internal/DoorbirdConfiguration.java:
```diff
@@ -0,0 +1,65 @@
+    /**
+     * Rate at which image channel will be updated
+     */
+    public Integer imageRefreshRate;
```
Since this parameter (and a few other ones) are optional, I am testing for the value of null to know whether or not the config parameter is set. Is there another way that's more preferred?

J-N-K replied:
I would set a default to 0 and check `>0` instead of `==null`.

mhilbush replied:
I understand your perspective, and that will certainly work. But, from a UI perspective, doesn't it make more sense for the field to be blank, rather than 0? A blank field would tend to suggest the feature (e.g. periodic image refresh) is unused. It just feels a bit more intuitive to me. WDYT?

J-N-K replied:
Empty will work also. We already have (at least) nikohomecontrol, sonyaudio, onkyo, dmx, modbus, velbus, neeo where "refresh=0" means "disable refresh".

mhilbush commented on bundles/org.openhab.binding.doorbird/src/main/java/org/openhab/binding/doorbird/internal/DoorbirdHandler.java:
```diff
@@ -0,0 +1,633 @@
+    private @Nullable ScheduledFuture<?> doorbellOffJob;
+    private @Nullable ScheduledFuture<?> motionOffJob;
+
+    private @Nullable String doorbirdId;
```
No strong reason. Since I was testing for the presence of these config parameters in `initialize`, it seemed to make sense to save the values for use later. Would you prefer I just save `config`, then reference the parameters using that?

J-N-K replied:
Yes. If you do that, you can also simplify e.g. `startImageRefreshJob`, no need to use `getConfigAs` again. You can use `config.imageRefreshRate` then.

mhilbush replied:
Ok, then that's what I'll do. And, you're right, it does simplify a number of methods. 😉 

Since `getConfigAs` is defined as `@NonNull`, and `config` is set in the `initialize` method only, is this an appropriate use of `@NonNullByDefault({})` in the declaration?

```
private @NonNullByDefault({}) DoorbirdConfiguration config;
```

mhilbush commented on bundles/org.openhab.binding.doorbird/src/main/java/org/openhab/binding/doorbird/internal/DoorbirdHandler.java:
```diff
@@ -0,0 +1,633 @@
+
+    @Override
+    public void initialize() {
+        logger.debug("Handler initializing for {}", getThing().getUID());
```
Removed

mhilbush commented on bundles/org.openhab.binding.doorbird/src/main/java/org/openhab/binding/doorbird/internal/DoorbirdHandler.java:
```diff
@@ -0,0 +1,633 @@
+        this.timeZoneProvider = timeZoneProvider;
+        this.httpClient = httpClient;
+        udpListener = new DoorbirdUdpListener(this);
+        logger.debug("Handler created for {}", thing.getUID());
```
Removed

mhilbush commented on bundles/org.openhab.binding.doorbird/src/main/java/org/openhab/binding/doorbird/internal/DoorbirdHandler.java:
```diff
@@ -0,0 +1,633 @@
+
+    @Override
+    public void dispose() {
+        logger.debug("Handler disposing for {}", getThing().getUID());
```
Removed

mhilbush commented on bundles/org.openhab.binding.doorbird/src/main/java/org/openhab/binding/doorbird/internal/DoorbirdHandler.java:
```diff
@@ -0,0 +1,633 @@
+    }
+
+    // Callback used by listener to update doorbell channel
+    public synchronized void updateDoorbellChannel(long timestamp) {
```
I wanted to ensure there were no overlapping calls to retrieve the image from the device. Is there a more preferred way to do this? Would it make more sense to instead synchronize on `downloadImage`?

J-N-K replied:
Sounds better.

mhilbush replied:
Fixed. And, in fact, `downloadImage` already was synchronized. So the other synchronized methods were unnecessary.

mhilbush commented on bundles/org.openhab.binding.doorbird/src/main/feature/feature.xml:
```diff
@@ -0,0 +1,11 @@
+  <feature name="openhab-binding-doorbird" description="Doorbird Binding" version="${project.version}">
+    <feature>openhab-runtime-base</feature>
+    <bundle start-level="80">mvn:org.openhab.addons.bundles/org.openhab.binding.doorbird/${project.version}</bundle>
+    <bundle dependency="true">wrap:mvn:com.goterl.lazycode/lazysodium-java/4.0.1$Bundle-Name=LazySodium-Java&amp;Bundle-SymbolicName=com.goterl.lazycode.lazysodium-java&amp;Bundle-Version=4.0.0</bundle>
```
Ok. So when I submit the PR for that repo, I just change this line to this?

```
<bundle dependency="true">mvn:com.goterl.lazycode/lazysodium-java/4.0.1</bundle>
```

mhilbush replied:
Hmm. After making these changes to feature.xml and pom.xml, you can't really build the binding jar until the PR for `openhab-osgiify` is merged. Am I missing something?

J-N-K replied:
Use the full name `org.openhab.osgiify/com.goterl.lazycode.lazysodium-java/4.0.1`. After you build the bundle to your local m2-repo, it should work.

mhilbush replied:
Got it. Now if I can figure out how to do the build in the `openhab-osgiify` repo...

J-N-K replied:
see my reply over there

mhilbush replied:
Built my osgiified version of the LazySodium and the binding now builds again. Thanks for your help on this!

I'll work on the remaining issues then ping you when I push the changes.

mhilbush commented:
@J-N-K Thanks for the review!

I've addressed your comments, and asked for a few clarifications. I'll push a commit with all the changes once we've worked through the clarifications.

In the meantime, I'll work on the PR for the `openhab-osgiify` repo.

J-N-K commented on bundles/org.openhab.binding.doorbird/src/main/feature/feature.xml:
```diff
@@ -0,0 +1,11 @@
+    <feature>openhab-runtime-base</feature>
+    <bundle start-level="80">mvn:org.openhab.addons.bundles/org.openhab.binding.doorbird/${project.version}</bundle>
+    <bundle dependency="true">mvn:org.openhab.osgiify/com.goterl.lazycode.lazysodium-java/4.0.1</bundle>
+    <bundle dependency="true">mvn:net.java.dev.jna/jna/5.4.0</bundle>
```
better: change the compile dependency to 5.4.0 if that works.

mhilbush replied:
Ack! I changed it in the pom but not here. I'll push a change to make them consistent.

I'll see if 5.4.0 works and, if so, I'll use that. I know LazySodium uses 5.3.0, so I might need to stick with that version. I'll test then push a change with one or the other.

J-N-K replied:
I had a look at the changelog and there were only additions and bugfixes, no breaking changes. So if should work.

mhilbush replied:
I changed to 5.4.0 and it's working with no issues.

cweitkamp commented on bundles/org.openhab.binding.doorbird/src/main/java/org/openhab/binding/doorbird/internal/listener/DoorbirdUdpListener.java:
```diff
@@ -0,0 +1,152 @@
+    private final DoorbirdEvent event = new DoorbirdEvent();
+
+    // Used for callbacks to handler
+    private DoorbirdHandler thingHandler;
```
```suggestion
    private final DoorbirdHandler thingHandler;
```

mhilbush replied:
Done

cweitkamp commented on bundles/org.openhab.binding.doorbird/src/main/java/org/openhab/binding/doorbird/internal/profile/DoorbirdProfileFactory.java:
```diff
@@ -0,0 +1,67 @@
+    private static final Set<ProfileType> SUPPORTED_PROFILE_TYPES = Collections
+            .singleton(DoorbirdProfiles.DOORBELL_COMMAND_TYPE);
+
+    private static final Set<ProfileTypeUID> SUPPORTED_PROFILE_TYPE_UIDS = Stream
```
You can use `Collections.singleton()` in this case too.

mhilbush replied:
Done

cweitkamp commented on bundles/org.openhab.binding.doorbird/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,224 @@
+		<description>Hangup a SIP call</description>
+		<state pattern="%s"></state>
+	</channel-type>
+	<channel-type id="restart">
```
Can we model this - the restart channel - as a thing action?

mhilbush replied:
Yes, it certainly could be modeled that way. I looked at the VolvoOnCall binding, and it looks pretty straightforward to set it up as an action. I'm only wondering if it makes sense to do that if `restart` is the only action supported by the binding. WDYT?

cweitkamp replied:
Yes, why not?  It can be a simple action without any parameters.

And maybe in the future there will be more :wink:. Point is I would like to push some of the new features in OH and thing actions can be used with the new rule engine (NGRE).

mhilbush replied:
Ok, that seems reasonable. I'll change this to a thing action.

mhilbush replied:
> And maybe in the future there will be more

I converted the `restart` and `sipHangup` channels to thing actions. I put the action package as a subpackage of `doorbird` (versus `doorbird.internal`). I saw bindings do it both ways, so I'm not sure which is preferred.

cweitkamp replied:
Looks good. :+1: I'd tend to internal.

mhilbush replied:
Hmm. When I move the `action` package under `internal`, I'm not able to invoke the action from within a rule. the rule throws an error that says
```
'sipHangup' is not a member of 'org.eclipse.smarthome.core.thing.binding.ThingActions'
```

Is there something else I need to do to cause the actions to be available in a rule? Does the package need to be exported somehow?

J-N-K replied:
Since packages that match `*.internal.*` are not exported, I don't think this will work.

mhilbush replied:
I was wondering about that. But there are bindings that have their `action` package under `internal`. See below. Then those actions should not work either?

https://github.com/openhab/openhab2-addons/blob/9c2629cd0bdd53276a9e3439bb52cd83f278627c/bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/action/VolvoOnCallActions.java

https://github.com/openhab/openhab2-addons/blob/86191c59572ee16852edc77170de0829f675a0b0/bundles/org.openhab.binding.pushbullet/src/main/java/org/openhab/binding/pushbullet/internal/action/PushbulletActions.java

https://github.com/openhab/openhab2-addons/blob/b3125023fe01fd1474e25e1eb68227e9d40ad289/bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/action/LightActions.java

cweitkamp replied:
IIRC there is a spacial handling inside the core framework to expose `ThingActions` towards DSL (see https://github.com/openhab/openhab-core/blob/20faf1c63b02f150501f81995db39e7b0d5d1bb3/bundles/org.openhab.core.model.script/src/org/eclipse/smarthome/model/script/scoping/ScriptImplicitlyImportedTypes.java#L129-L134). Please correct me if I am wrong.

The problem we might face here is that the import will be done once during start if the DSL rule engine. Meaning if you add a binding later during runtime an new action might not be available immediately. The rule engine has to be restarted.

mhilbush replied:
Given that limitation, and as there are other bindings that do it this way, then I would argue to leave the package where it is (i.e. as a subpackage to `doorbird`). Ok?

cweitkamp replied:
Yes, fine for me.

cweitkamp commented on bundles/org.openhab.binding.doorbird/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,224 @@
+        <kind>trigger</kind>
+		<label>Motion</label>
+		<description>Motion detected</description>
+        <event>
```
Something looks wrong with formatting.

mhilbush replied:
Corrected

cweitkamp commented on bundles/org.openhab.binding.doorbird/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,224 @@
+		<label>Doorbird D210x</label>
+		<description>Doorbird doorbell model D210x</description>
+		<channels>
+			<channel id="doorbell" typeId="doorbell" />
```
We can use system default channel type `rawbutton` for this channel.

mhilbush replied:
@cweitkamp I'm a little confused about how to change this. Do I:

- define this channel to have type `system.rawbutton` (and delete the doorbell channel type further down in this file), OR

- do I leave it as a trigger channel, and reference the the `rawbutton-on-off-switch` profile when defining the item for the `doorbell` channel?

cweitkamp replied:
I would do it like this:

```suggestion
			<channel id="doorbell" typeId="system.rawbutton" />
```

You can remove channel type `doorbell` ([line 59](https://github.com/openhab/openhab2-addons/pull/5857/files#diff-f44bb87b35e8aa02edaf40d8ec8e6643R59)) from this file.

mhilbush replied:
Done

cweitkamp commented on bundles/org.openhab.binding.doorbird/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,224 @@
+			<channel id="doorbell" typeId="doorbell" />
+			<channel id="doorbellTimestamp" typeId="doorbellTimestamp" />
+			<channel id="doorbellImage" typeId="doorbellImage" />
+			<channel id="motion" typeId="motion" />
```
We can use system default channel type `trigger` for this channel.

cweitkamp replied:
Outdated: see https://github.com/openhab/openhab2-addons/pull/5857/files#r331783379

cweitkamp replied:
```suggestion
			<channel id="motion" typeId="system.motion" />
```

You can remove channel type `motion` ([line 82](https://github.com/openhab/openhab2-addons)) from this file./pull/5857/files#diff-f44bb87b35e8aa02edaf40d8ec8e6643R82

cweitkamp commented on bundles/org.openhab.binding.doorbird/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,224 @@
+		<description>Historical image for motion sensor</description>
+		<state readOnly="true"></state>
+	</channel-type>
+	<channel-type id="getDoorbellMontage" advanced="false">
```
Missing in the docs? Can it be a thing action? Or `RefreshType.REFRESH`?

mhilbush replied:
Removed channel and will use REFRESH on the associated image channel.

cweitkamp commented on bundles/org.openhab.binding.doorbird/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,224 @@
+		<description>Montage of multiple doorbell history images</description>
+		<state readOnly="true"></state>
+	</channel-type>
+	<channel-type id="getMotionMontage" advanced="false">
```
Missing in the docs? Can it be a thing action? Or `RefreshType.REFRESH`?

mhilbush replied:
Removed channel and will use REFRESH on the associated image channel.

cweitkamp commented on bundles/org.openhab.binding.doorbird/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,224 @@
+		<description>Energize opendoor / alarm output relay 2</description>
+		<state pattern="%s"></state>
+	</channel-type>
+	<channel-type id="getImage">
```
What is the purpose of this channel? To update the ìmage` channel`? Why not force an update of the image channel by sending an `RefreshType.REFRESH` to it?

cweitkamp commented on bundles/org.openhab.binding.doorbird/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,224 @@
+		<label>Doorbird D101</label>
+		<description>Doorbird doorbell model D101</description>
+		<channels>
+			<channel id="doorbell" typeId="doorbell" />
```
We can use system default channel type `rawbutton` for this channel.

mhilbush replied:
Done

cweitkamp commented on bundles/org.openhab.binding.doorbird/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,224 @@
+			<channel id="doorbell" typeId="doorbell" />
+			<channel id="doorbellTimestamp" typeId="doorbellTimestamp" />
+			<channel id="doorbellImage" typeId="doorbellImage" />
+			<channel id="motion" typeId="motion" />
```
We can use system default channel type `trigger` for this channel.

cweitkamp replied:
Outdated: see https://github.com/openhab/openhab2-addons/pull/5857/files#r331783379

cweitkamp commented on bundles/org.openhab.binding.doorbird/src/main/java/org/openhab/binding/doorbird/internal/DoorbirdHandler.java:
```diff
@@ -0,0 +1,622 @@
+    private byte @Nullable [] convertImageToByteArray(BufferedImage image) {
+        byte[] data = null;
+        try {
+            ByteArrayOutputStream out = new ByteArrayOutputStream();
```
You can use [try-with-resources](https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html) statement here.

mhilbush replied:
Done

cweitkamp commented on bundles/org.openhab.binding.doorbird/src/main/java/org/openhab/binding/doorbird/internal/DoorbirdHandler.java:
```diff
@@ -0,0 +1,622 @@
+    // Get a dedicated threadpool for the long-running listener thread
+    private final ScheduledExecutorService doorbirdScheduler = ThreadPoolManager.getScheduledPool("doorbirdHandler");
+    private @Nullable ScheduledFuture<?> listenerJob;
+    private DoorbirdUdpListener udpListener;
```
```suggestion
    private final DoorbirdUdpListener udpListener;
```

mhilbush replied:
Done

cweitkamp commented on bundles/org.openhab.binding.doorbird/src/main/java/org/openhab/binding/doorbird/internal/DoorbirdHandler.java:
```diff
@@ -0,0 +1,622 @@
+
+    private @Nullable String authorization;
+
+    private TimeZoneProvider timeZoneProvider;
```
```suggestion
    private final TimeZoneProvider timeZoneProvider;
```

mhilbush replied:
Done

cweitkamp commented on bundles/org.openhab.binding.doorbird/src/main/java/org/openhab/binding/doorbird/internal/DoorbirdHandler.java:
```diff
@@ -0,0 +1,622 @@
+    private @Nullable String authorization;
+
+    private TimeZoneProvider timeZoneProvider;
+    private HttpClient httpClient;
```
```suggestion
    private final HttpClient httpClient;
```

mhilbush replied:
Done

cweitkamp commented on bundles/org.openhab.binding.doorbird/src/main/java/org/openhab/binding/doorbird/internal/DoorbirdHandler.java:
```diff
@@ -0,0 +1,622 @@
+    private @Nullable ScheduledFuture<?> doorbellOffJob;
+    private @Nullable ScheduledFuture<?> motionOffJob;
+
+    private @NonNullByDefault({}) DoorbirdConfiguration config;
```
This has to be annotated with `@Nullable` too.

```suggestion
    private @Nullable DoorbirdConfiguration config;
```

J-N-K replied:
My understanding of the last discussion (which unfortunately I can‘t find ATM) was that „I know this to be NonNull before use by design“ (which is IMO the case here) also qualifies for `@NonNullByDefault({})`.

cweitkamp replied:
Let me quote something from our sorely missed OHC maintainer (see https://github.com/openhab/openhab-core/pull/910#discussion_r300031597):

> We introduced the usage of `@NonNullByDefault({})` for variables we know that they are not null but this cannot be detected by the tooling. One common example has been references of components that are injected by the declarative service implementation on activation.
For this use case there is no need anymore because every mandatory reference can now be set by using constructor injection.
For optional references, we should already use `@Nullable` because they can be valid null.
Another use case are members that are injected by reflection because of specific deserialization (e.g. Gson).
>
> Why do we need it in this case.
> Wouldn't it be correct to mark these variable `@Nullable`?

mhilbush replied:
I saw other bindings do this, so I thought it was ok. For example, 😉 
https://github.com/openhab/openhab2-addons/blob/9d5b40bff36242cd17757e219b4f660831b9ca8a/bundles/org.openhab.binding.darksky/src/main/java/org/openhab/binding/darksky/internal/handler/DarkSkyAPIHandler.java#L64

I'll do whatever you guys want. But, won't marking it nullable require null checks wherever its referenced?

cweitkamp replied:
Touché! Fair enough. I used it in the DarkSky binding and elsewhere for the same reason like you. Because it worked and no one complained about it. But rules and common usage may change over time ...

> But, won't marking it nullable require null checks wherever its referenced?

Yes, it does. And that is the way to go because it could be null and we are working in an multi threaded environment where such things could result in NPEs.

Another option could be to set the `config` property in line 90 a non null value:

```suggestion
    private DoorbirdConfiguration config = new DoorbirdConfiguration();
```

J-N-K replied:
I still don‘t understand how this should happen. The framework should never call handler-methods on an uninitialized thing. And initialize() guarantees that this is NoNull when it returns. There is no code that re-sets this to null.

mhilbush replied:
Right, `config` is set in one and only one place (i.e. in `initialize`). And, `getConfigAs` is not defined as `@Nullable` in `BaseThingHandler`, so it's guaranteed to return non-null. I don't see where `config` can ever be null once the thing is initialized.

cweitkamp replied:
Alright. I did not want to go beyond the scope and screw up everything. I agree, that we can be sure that `config` cannot be null. It is up to you if you want to change it or leave it like it is.

cweitkamp commented on bundles/org.openhab.binding.doorbird/src/main/java/org/openhab/binding/doorbird/internal/DoorbirdHandler.java:
```diff
@@ -0,0 +1,622 @@
+    }
+
+    private void handleLight(Command command) {
+        if (command instanceof OnOffType && command.equals(OnOffType.ON)) {
```
Is it not possible to turn off the light?

mhilbush replied:
No. The device turns it off automatically after a defined amount of time.

cweitkamp commented on bundles/org.openhab.binding.doorbird/src/main/java/org/openhab/binding/doorbird/internal/profile/DoorbirdProfileFactory.java:
```diff
@@ -0,0 +1,67 @@
+
+    @Nullable
+    @Override
+    public Profile createProfile(ProfileTypeUID profileTypeUID, ProfileCallback callback, ProfileContext context) {
```
Minor: wrong location of `@Nullable`:

```suggestion
    public @Nullable Profile createProfile(ProfileTypeUID profileTypeUID, ProfileCallback callback, ProfileContext context) {
```

mhilbush replied:
Done

mhilbush commented:
@cweitkamp Thanks for your feedback. I'll work on these, then ping you when I'm complete, or with any follow up questions.

Hilbrand commented on bundles/org.openhab.binding.doorbird/README.md:
```diff
@@ -0,0 +1,239 @@
+
+| Parameter | Required/Optional | Description |
+|-----------|-------------------|-------------|
+| ID | Required | The ID number of the Doorbird device. This is usually a single digit (e.g. 1). |
```
You can easily format the tables with online tools like: http://markdowntable.com/

mhilbush replied:
See the discussion with @J-N-K [here](https://github.com/openhab/openhab2-addons/pull/5857#discussion_r316365507). I was trying avoid messing with alignment, which can be quite tedious. However, the link you provided makes it easy, so I'll replace what's here with the column-aligned version of the table.

mhilbush replied:
Done

mhilbush commented on bundles/org.openhab.binding.doorbird/src/main/java/org/openhab/binding/doorbird/internal/DoorbirdHandler.java:
```diff
@@ -0,0 +1,622 @@
+    }
+
+    private DateTimeType getLocalDateTimeType(long dateTimeSeconds) {
+        Instant instant = Instant.ofEpochMilli(dateTimeSeconds * 1000);
```
Thanks for the review @Hilbrand.

`dateTimeSeconds` is in seconds, so the x1000 is needed in the call to `ofEpochMilli`.

Hilbrand replied:
Ah ok. Didn't I used the wrong logic :smile: I see there is also `Instant.ofEpochSecond`

mhilbush replied:
Thanks! I missed the `ofEpochSecond`. I'll change to use that.

mhilbush replied:
Done

mhilbush commented on bundles/org.openhab.binding.doorbird/src/main/java/org/openhab/binding/doorbird/internal/DoorbirdBindingConstants.java:
```diff
@@ -0,0 +1,70 @@
+    public static final String CHANNEL_RESTART = "restart";
+
+    // List of trigger events
+    public static final String EVENT_PRESSED = "PRESSED";
```
Done. Also CommonTriggerEvents.RELEASED.

mhilbush commented on bundles/org.openhab.binding.doorbird/src/main/java/org/openhab/binding/doorbird/internal/DoorbirdBindingConstants.java:
```diff
@@ -0,0 +1,70 @@
+    public static final String EVENT_PRESSED = "PRESSED";
+    public static final String EVENT_RELEASED = "RELEASED";
+    public static final String EVENT_TRIGGERED = "TRIGGERED";
+    public static final String EVENT_UNTRIGGERED = "UNTRIGGERED";
```
It's used in the `startMotionOffJob` method.

cweitkamp replied:
Sry for my silly question: But what does it mean and what is a use-case for it? I mean an event can be triggered. That is clear. Can it be "untriggered"? Is the event happening over a period of time?

mhilbush replied:
> Sry for my silly question

There are no silly questions... :wink:

Most motion sensors with which I'm familiar are in either a triggered (ON) state, or an untriggered (OFF) state. I'm most familiar with zwave motion detectors where they are in an ON state when motion is detected, then after a period of time (sometimes configurable) they change back to an OFF state.

I'm trying to mirror that behavior so that the item referencing the switch profile can be either ON when motion is detected, then become OFF after a period of time. I was using the `UNTRIGGERED` event to cause the transition to OFF after a user-defined number of seconds.

The Doorbird motion sensor is quite sensitive, so it tends to detect motion quite easily (almost to the point of being unusable). Therefore, the specific use case for the Doorbird is give the user the ability to define an interval during which Doorbird motion detection events will not be propagated to the user through the item that references the switch profile.

cweitkamp replied:
Why don't you use a basic Switch channel for it at all ([system default channel type](https://www.openhab.org/docs/developer/bindings/thing-xml.html#system-state-channel-types) `motion`)?

mhilbush replied:
Well... The doorbell and motion channels originally were defined as switches. I later changed them to be triggers, but wanted to preserve the ON/OFF functionality. Maybe I should just change them both back to switches. It will make everything much simpler. WDYT?

cweitkamp replied:
I am completely okay with `doorbell` channel to be a trigger but imo `motion` should be a Switch.

mhilbush commented on bundles/org.openhab.binding.doorbird/src/main/java/org/openhab/binding/doorbird/internal/DoorbirdHandler.java:
```diff
@@ -0,0 +1,622 @@
+            errorMsg = "TimeoutException: Call to Doorbird API timed out";
+        } catch (ExecutionException e) {
+            errorMsg = String.format("ExecutionException: %s", e.getMessage());
+        } catch (InterruptedException e) {
```
Done

mhilbush commented on bundles/org.openhab.binding.doorbird/src/main/java/org/openhab/binding/doorbird/internal/DoorbirdHandler.java:
```diff
@@ -0,0 +1,622 @@
+    private byte @Nullable [] convertImageToByteArray(BufferedImage image) {
+        byte[] data = null;
+        try {
+            ByteArrayOutputStream out = new ByteArrayOutputStream();
```
Done

mhilbush commented on bundles/org.openhab.binding.doorbird/README.md:
```diff
@@ -0,0 +1,239 @@
+
+## Profiles
+
+A Switch profile is available for the doorbell and motion channels that will cause ON/OFF 
```
@cweitkamp Regarding your point #1 above, I'm not familiar with the `system.rawbutton` channel type nor the `rawbutton-toggle-switch` profile.

If I change the channel type of my `doorbell` channel to `system.rawbutton`, does that mean I can completely delete the profile package in my source tree? And, other than changing the channel type to `system.rawbutton`, is there anything else I need to do to use the `rawbutton-toggle-switch` profile?

mhilbush replied:
And, in the case of your point 2 above, we would 

- add `TRIGGERED` and `UNTRIGGERED` to [this file](https://github.com/openhab/openhab-core/blob/48d873a32ba0901a9b58f04c541f0f858fbfcd51/bundles/org.openhab.core.thing/src/main/java/org/eclipse/smarthome/core/thing/CommonTriggerEvents.java#L27), and

- add a new profile to [this file](https://github.com/openhab/openhab-core/blob/8d5d78937116741cc10313a47b057414f2708a15/bundles/org.openhab.core.thing/src/main/java/org/eclipse/smarthome/core/thing/profiles/SystemProfiles.java) (although I'm not exactly sure how this would be defined).

Anything else?

cweitkamp replied:
Let's have a look into the avmfritz binding how it is working:

Definition:
https://github.com/openhab/openhab2-addons/blob/f39c7b46e1837b9682eb32ca58c78eebebe47629/bundles/org.openhab.binding.avmfritz/src/main/resources/ESH-INF/thing/thing-types.xml#L213-L216

The handler uses the `triggerChannel(channelUID, event)` method to trigger a `CommonTriggerEvents.PRESSED` or `CommonTriggerEvents.RELEASED` event. The `rawbutton-toggle-switch` profiles is listening to the first event and toggles a switch.

A typical item definition will look like this:
```
   channel="avmfritz:HAN_FUN_SWITCH:<BRIDGE_ID>:<THING_ID>:press"[profile="rawbutton-toggle-switch"]
```

That's it. I would add a note and link to the profile section in the [items](https://www.openhab.org/docs/configuration/items.html#profiles) page of the official docs.

> does that mean I can completely delete the profile package in my source tree? 
Yes,  you can remove the profile package from your binding.

Now I can see the difference. You need a profile which act a little bit different. `PRESSED` => `ON` and `RELEASED` => `OFF`. Not a toggle on `PRESSED`. Another potential system default profile.

> add TRIGGERED and UNTRIGGERED to this file
Correct. But as mentioned in https://github.com/openhab/openhab2-addons/pull/5857#discussion_r331658260 I'm not sure about the meaning of `UNTRIGGERED`.

> add a new profile to this file (although I'm not exactly sure how this would be defined).
Definition has to be done in the linked file. You can find the factory and available handlers in the [internal package](https://github.com/openhab/openhab-core/tree/8d5d78937116741cc10313a47b057414f2708a15/bundles/org.openhab.core.thing/src/main/java/org/eclipse/smarthome/core/thing/internal/profiles).

mhilbush replied:
Ok, that all makes sense. I'll take a whack at the changes necessary to support this. I may have a few more questions as I get into it. I'm pretty comfortable working on bindings, but this is probably on the outer limit of my skill level.

mhilbush replied:
I've converted the `motion` channel back to a switch.

So, now w.r.t. the `doorbell` channel, I guess I need to create a new profile in core that will handle the `PRESSED` -> `ON` and `RELEASED` -> `OFF` switch functionality.

My only misgiving with this is that anyone who's not on a recent snapshot release won't be able to use the binding (and there are quite a few people using it), as the new profile won't be available.

cweitkamp replied:
Yes, but a rule to imitate a similar behavior shouldn't be too complex as a replacement. Isn't it?

mhilbush replied:
Ok, @cweitkamp, I pushed an openhab-core PR to add the switch ON/OFF profile functionality. Please have a look and let me know if that's what you were expecting.

mhilbush replied:
Done

mhilbush commented:
@cweitkamp I *think* I've now addressed all the outstanding review comments from you, @J-N-K and @Hilbrand.

mhilbush commented on bundles/org.openhab.binding.doorbird/src/main/java/org/openhab/binding/doorbird/internal/listener/DoorbirdUdpListener.java:
```diff
@@ -0,0 +1,162 @@
+                // Lengths are different, therefore not a dup
+                if (Arrays.equals(lastData, Arrays.copyOf(packet.getData(), packet.getLength()))) {
+                    // Packet must be received within 750 ms of previous packet to be consider a duplicate
+                    if ((System.currentTimeMillis() - lastDataTime) < 750) {
```
Good catch. Done.

mhilbush commented on bundles/org.openhab.binding.doorbird/src/main/java/org/openhab/binding/doorbird/internal/listener/DoorbirdUdpListener.java:
```diff
@@ -0,0 +1,162 @@
+
+    private boolean isDuplicate(DatagramPacket packet) {
+        boolean packetIsDuplicate = false;
+        if (lastData != null) {
```
Ok, done.

mhilbush commented on bundles/org.openhab.binding.doorbird/src/main/java/org/openhab/binding/doorbird/internal/DoorbirdHandler.java:
```diff
@@ -0,0 +1,615 @@
+ */
+@NonNullByDefault
+public class DoorbirdHandler extends BaseThingHandler {
+    private static final long API_REQUEST_TIMEOUT = 3L;
```
Done

mhilbush commented on bundles/org.openhab.binding.doorbird/src/main/java/org/openhab/binding/doorbird/internal/DoorbirdHandler.java:
```diff
@@ -0,0 +1,615 @@
+@NonNullByDefault
+public class DoorbirdHandler extends BaseThingHandler {
+    private static final long API_REQUEST_TIMEOUT = 3L;
+    private static final long MONTAGE_UPDATE_DELAY = 5L;
```
Done

mhilbush commented:
@J-N-K Do you need to do any additional reviews of this? With @cweitkamp and @Hilbrand approvals, is this ready to merge? I don't mean to be a pest, but there's a new device I'd like to add (A1081 controller requires new thing type, channels, handler, etc.), and I'd prefer to start work on that knowing I won't be making further changes here. Thanks!

J-N-K commented:
Thanks.

mhilbush commented:
Thanks guys for your help. Very much appreciated!
